pkg/sqlutil/pg/connection.go

```diff
@@ -0,0 +1,118 @@
+package pg
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"time"
+
+	"github.com/XSAM/otelsql"
+	"github.com/google/uuid"
+	"github.com/jackc/pgx/v4"
+	"github.com/jackc/pgx/v4/stdlib"
+	"github.com/jmoiron/sqlx"
+	"github.com/scylladb/go-reflectx"
+	"go.opentelemetry.io/otel"
+	semconv "go.opentelemetry.io/otel/semconv/v1.4.0"
+
+	// need to make sure pgx driver is registered before opening connection
+	_ "github.com/jackc/pgx/v4/stdlib"
+)
+
+// NOTE: This is the default level in Postgres anyway, we just make it
+// explicit here
+const defaultIsolation = sql.LevelReadCommitted
+
+// DialectName is a compiler enforced type used that maps to database dialect names
+type DialectName string
+
+const (
+	// DialectPostgres represents the postgres dialect.
+	DialectPostgres DialectName = "pgx"
+	// DialectTransactionWrappedPostgres is useful for tests.
+	// When the connection is opened, it starts a transaction and all
+	// operations performed on the DB will be within that transaction.
+	DialectTransactionWrappedPostgres DialectName = "txdb"
+)
+
+type ConnectionConfig struct {
+	DefaultIdleInTxSessionTimeout time.Duration
+	DefaultLockTimeout            time.Duration
+	MaxOpenConns                  int
+	MaxIdleConns                  int
+}
+
+func (config ConnectionConfig) NewDB(uri string, dialect DialectName) (*sqlx.DB, error) {
+	opts := []otelsql.Option{otelsql.WithAttributes(semconv.DBSystemPostgreSQL),
+		otelsql.WithTracerProvider(otel.GetTracerProvider()),
+		otelsql.WithSQLCommenter(true),
+		otelsql.WithSpanOptions(otelsql.SpanOptions{
+			OmitConnResetSession: true,
+			OmitConnPrepare:      true,
+			OmitRows:             true,
+			OmitConnectorConnect: true,
+			OmitConnQuery:        false,
+		})}
+
+	// Set default connection options
+	lockTimeout := config.DefaultLockTimeout.Milliseconds()
+	idleInTxSessionTimeout := config.DefaultIdleInTxSessionTimeout.Milliseconds()
+	stmt := fmt.Sprintf(`SET TIME ZONE 'UTC'; SET lock_timeout = %d; SET idle_in_transaction_session_timeout = %d; SET default_transaction_isolation = %q`,
+		lockTimeout, idleInTxSessionTimeout, defaultIsolation.String())
+
+	var sqldb *sql.DB
+	if dialect == DialectTransactionWrappedPostgres {
```

Comment:
Alternatively we could have this:
- trigger based on a switch in `ConnectionConfig`, so that we don't have to make any assumptions about specific dialects from here.
- split out as a separate methods, perhaps `Open` and `OpenDB` to align with `sql` and `sqlx`

Both cases burden the caller slightly, but trivially and with little to no risk. FWIW the config version would be less risky, since the default would be to use a connector, and there is no decision to be made unless you are looking for it.